_posts/2022-09-08-storage-write-removal-before-conditional-termination.md

```diff
@@ -0,0 +1,168 @@
+- In all code paths after the initial write, a subsequent write overwrites the written value.
+- The transaction unconditionally reverts after the write.
+
+So similarly to above, the general pattern is:
+1. a storage write (potentially within complex control flow), then
+2. potentially unrelated code, and finally 
+3. any continuing control flow path either overwrites the storage write in (1) with a different value or reverts.
+
+The bug is due to the treatment of certain function calls in (2).
+
+If a function call is performed between (1) and (3), the optimizer has to consider the control flow behaviour of the function call, e.g. whether control flow may continue beyond the call, whether the called function always reverts or always successfully terminates. However, in the case that the control flow conditionally can continue after the call to a function, but the function call may also terminate using the ``return(...)`` or ``stop()`` statements (see examples below), the optimizer incorrectly still behaved, as if the control flow *always* continued after the function.
+
+Hence the write in (1) could be removed, even though (2) can in fact terminate and is supposed to retain the storage write in (1).
+
+So on the Yul level the issue manifests if the following sequence occurs within a Yul block during optimization:
+
+1. A storage write occurs (e.g. ``sstore(0, 1)``, but the write may also occur within complex control flow like a single ``switch`` case).
+2. A user-defined function is called and this function both has a control flow path that terminates the transaction and a control flow path out of the function.
+3. Any continuing control flow path does either of the following:
+    - It overwrites the storage slot written in (1), s.t. the storage write in (1) becomes redundant for this path (e.g. ``sstore(0, 2)``).
+    - The control flow unconditionally reverts (e.g. ``revert(0, 0)``)
+
+In this situation, the optimizer will treat the storage write in (1) as redundant and remove it, despite the fact that the called Yul function may successfully terminate the EVM call, in which case this storage write is in fact not redundant at all.
+
+The following is a full example of an affected Yul snippet:
+
+```yul
+{
+	function f() {
+		if gt(calldatasize(), 4) { leave }
+		return(0, 0)
+	}
+	sstore(0, 1) // This sstore will incorrectly be removed.
+	f()
+	sstore(0, 2)
+}
+```
+
+In the absence of the bug, the first ``sstore`` would have been be marked as required by the optimizer upon encountering the call to ``f()``, i.e. it should not be considered for removal regardless of what happens afterwards, since the control flow path of ``f()`` reaching the ``return(0, 0)`` already relies on the ``sstore`` happening.
```

Comment:
```suggestion
In the absence of the bug, the first ``sstore`` would have been marked as required by the optimizer upon encountering the call to ``f()``, i.e. it should not be considered for removal regardless of what happens afterwards, since the control flow path of ``f()`` reaching the ``return(0, 0)`` already relies on the ``sstore`` happening.
```